counsel.el

```diff
@@ -1892,6 +1893,32 @@ Skip some dotfiles unless `ivy-text' requires them."
 (defvar counsel-find-file-speedup-remote t
   "Speed up opening remote files by disabling `find-file-hook' for them.")
 
+(defun counsel-read-env ()
+  "Read a file path environment variable and insert it into the
+minibuffer."
+  (interactive)
+  (if (equal ivy-text "")
+      (let ((enable-recursive-minibuffers t)
+            (old-last ivy-last))
+        (ivy-read "Env: "
+                  (cl-loop for pair in process-environment
+                           for (var val) = (split-string pair "=" t)
+                           if (and val (not (equal "" val)))
+                           if (file-exists-p
+                               (if (file-name-absolute-p val)
+                                   val
+                                 (setq val
+                                       (expand-file-name val ivy--directory))))
+                           collect (cons var val))
+                  :action (lambda (x)
+                            (ivy--reset-state (setq ivy-last old-last))
+                            (let ((path (cdr x)))
+                              (when (file-accessible-directory-p path)
+                                (setq path (file-name-as-directory path)))
+                              (insert (abbreviate-file-name path)))
+                            (ivy--cd-maybe))))
+    (insert "$")))
```

Comment:
This and the associated key binding couple this command with the <kbd>$</kbd> key. I think this feature would be less intrusive if it worked similarly to e.g. magic slash, which is controlled by a user option and works by reacting to entered text, rather than specific key bindings. Key bindings are for customisation, not for business logic.

Reply:
okay. I have applied your feedback. Perhaps you can help me with the two issues that I am now hitting because of the change. 

1. using `keyboard-quit` will now lead to an error and break the `post-command-hook` when reading env variables.
2. environment variables that contain a filename will no longer limit the candidates to match. For example `(setenv "foo" "~/.bashrc")` then reading the variable `foo` will display all the candidates in the home directory, even though `.bashrc` is selected.

Reply:
Sorry, I don't have the time to dig into this at the moment. Perhaps @abo-abo can help. if the old approach works, the least you could do is emulate `self-insert-command` in inserting the key that invoked the command, rather than hardcoding the dollar sign.

Reply:
It looks like the issue is that `post-command-hook` treats `keyboard-quit` as an error. So I don't think that putting a recursive minibuffer in the `post-command-hook` will work.

> the least you could do is emulate `self-insert-command` in inserting the key that invoked the command, rather than hardcoding the dollar sign.

I don't understand what you mean. The key is bound to `$` and the function inserts `$`.

Reply:
> I don't understand what you mean. The key is bound to `$` and the function inserts `$`.

But what if a user unbinds or rebinds the function to, say, <kbd>%</kbd>? Then the function should insert `%`, not `$`. IIRC this information is saved in `last-command-event` or similar.

Reply:
gotcha